### ISSUE TITLE
Detection of `install` arg for `configure_file` should print file and line number

### ISSUE THREAD

**Issue by ignatenkobrain** [mesonbuild/meson#5048]
```
WARNING: Project targetting '>=0.47' but tried to use feature introduced in '0.50.0': install arg in configure_file
```

This has happened to tracker... And obviously it was never updated to require 0.50.0 yet..

Not sure if this is something wrong with tracker or meson.

**Comment by nirbheek** [comment]
I found this:

```meson
configure_file(
    input: 'tracker-miner.pc.in',
    output: 'tracker-miner-@0@.pc'.format(tracker_api_version),
    configuration: conf,
    install: true,
    install_dir: join_paths(get_option('prefix'), get_option('libdir'), 'pkgconfig'))
```

So it's not a false positive. It should definitely print the file and line number, though. That's a real bug.

**Comment by mcatanzaro** [comment]
Why is it a bug? Epiphany uses this to install its search provider service and default bookmarks. It's always worked fine. The generated files get installed exactly where expected.

The feature also does not seem to be new in 0.50.0:

https://github.com/mesonbuild/meson/compare/0.50.0...master

So I'm confused.

**Comment by nirbheek** [comment]
This is a new feature. In the past the `install:` kwarg was just ignored and did nothing. Only the `install_dir:` kwarg was used. 

https://github.com/mesonbuild/meson/commit/95c1cdf7767d5b5bf4614651afeafa100f1de45c and https://github.com/mesonbuild/meson/issues/3983

**Comment by mcatanzaro** [comment]
Eh, OK... the compatibility message isn't really useful though, since there's no reason to require meson 0.50.0 other than to silence the warning message.

**Comment by nirbheek** [comment]
There is: if you're using f.ex., `install: get_option('installed-tests')`, your build files are broken unless you're using 0.50 or newer.

**Comment by mcatanzaro** [comment]
Eh, epiphany uses `install: true` so we're going to be bumping required meson version for no reason... not a big deal, anyway.

**Comment by nirbheek** [comment]
If the purpose is to get rid of the warning on newer versions of Epiphany, you can comment out that line and add a TODO 'uncomment when bumping meson req'.

**Comment by jpakkane** [comment]
I looked at adding this but it turned out to be a lot of work. The file and line number information is lost fairly often and the checker is called from many different places. For example, all Targets should be updated to have the file and line numbers where they are defined.

This is as far as I got when trying to fix this before giving up:

```diff --git a/mesonbuild/interpreter.py b/mesonbuild/interpreter.py
index 4b713ea0..5bd3a716 100644
--- a/mesonbuild/interpreter.py
+++ b/mesonbuild/interpreter.py
@@ -2926,16 +2926,16 @@ external dependencies (including libraries) must go to "dependencies".''')
                  mlog.bold(subproj_path), 'found:', mlog.green('YES'))
         return dep
 
-    def _handle_featurenew_dependencies(self, name):
+    def _handle_featurenew_dependencies(self, fname, lineno, name):
         'Do a feature check on dependencies used by this subproject'
         if name == 'mpi':
-            FeatureNew('MPI Dependency', '0.42.0').use(self.subproject)
+            FeatureNew('MPI Dependency', '0.42.0').use(fname, lineno, self.subproject)
         elif name == 'pcap':
-            FeatureNew('Pcap Dependency', '0.42.0').use(self.subproject)
+            FeatureNew('Pcap Dependency', '0.42.0').use(fname, lineno, self.subproject)
         elif name == 'vulkan':
-            FeatureNew('Vulkan Dependency', '0.42.0').use(self.subproject)
+            FeatureNew('Vulkan Dependency', '0.42.0').use(fname, lineno, self.subproject)
         elif name == 'libwmf':
-            FeatureNew('LibWMF Dependency', '0.44.0').use(self.subproject)
+            FeatureNew('LibWMF Dependency', '0.44.0').use(fname, lineno, self.subproject)
         elif name == 'openmp':
             FeatureNew('OpenMP Dependency', '0.46.0').use(self.subproject)
 
@@ -2953,7 +2953,7 @@ external dependencies (including libraries) must go to "dependencies".''')
         if not isinstance(not_found_message, str):
             raise InvalidArguments('The not_found_message must be a string.')
         try:
-            d = self.dependency_impl(name, display_name, kwargs)
+            d = self.dependency_impl(node, name, display_name, kwargs)
         except Exception:
             if not_found_message:
                 self.message_impl(not_found_message)
@@ -2962,7 +2962,7 @@ external dependencies (including libraries) must go to "dependencies".''')
             self.message_impl(not_found_message)
         return d
 
-    def dependency_impl(self, name, display_name, kwargs):
+    def dependency_impl(self, node, name, display_name, kwargs):
         disabled, required, feature = extract_required_kwarg(kwargs, self.subproject)
         if disabled:
             mlog.log('Dependency', mlog.bold(display_name), 'skipped: feature', mlog.bold(feature), 'disabled')
@@ -2998,7 +2998,9 @@ external dependencies (including libraries) must go to "dependencies".''')
         wrap_mode = self.coredata.get_builtin_option('wrap_mode')
         forcefallback = wrap_mode == WrapMode.forcefallback and has_fallback
         if name != '' and not forcefallback:
-            self._handle_featurenew_dependencies(name)
+            fname = os.path.join(node.subdir, 'meson.build')
+            lineno = node.lineno
+            self._handle_featurenew_dependencies(fname, lineno, name)
             kwargs['required'] = required and not has_fallback
             dep = dependencies.find_external_dependency(name, self.environment, kwargs)
             kwargs['required'] = required
@@ -3123,7 +3125,9 @@ external dependencies (including libraries) must go to "dependencies".''')
         if 'input' not in kwargs or 'output' not in kwargs:
             raise InterpreterException('Keyword arguments input and output must exist')
         if 'fallback' not in kwargs:
-            FeatureNew('Optional fallback in vcs_tag', '0.41.0').use(self.subproject)
+            fname = os.path.join(node.subdir, 'meson.build')
+            lineno = node.lineno
+            FeatureNew('Optional fallback in vcs_tag', '0.41.0').use(fname, lineno, self.subproject)
         fallback = kwargs.pop('fallback', self.project_version)
         if not isinstance(fallback, str):
             raise InterpreterException('Keyword argument fallback must be a string.')
diff --git a/mesonbuild/interpreterbase.py b/mesonbuild/interpreterbase.py
index 650d1e06..8e50f5d0 100644
--- a/mesonbuild/interpreterbase.py
+++ b/mesonbuild/interpreterbase.py
@@ -100,6 +100,15 @@ def _get_callee_args(wrapped_args, want_subproject=False):
     kwargs = kwargs if kwargs is not None else {}
     return s, node, args, kwargs, subproject
 
+def _fname_and_lineno_from_callee_args(callee_args):
+    node = callee_args[1]
+    if node is None:
+        return '', 0
+    fname = os.path.join(node.subdir, 'meson.build')
+    lineno = node.lineno
+    return fname, lineno
+
+
 def flatten(args):
     if isinstance(args, mparser.StringNode):
         return args.value
@@ -189,7 +198,7 @@ class FeatureCheckBase:
             return ''
         return mesonlib.project_meson_versions[subproject]
 
-    def use(self, subproject):
+    def use(self, fname, line_number, subproject):
         tv = self.get_target_version(subproject)
         # No target version
         if tv == '':
@@ -209,7 +218,7 @@ class FeatureCheckBase:
             # means we won't warn about a feature used in multiple places.
             return
         register[self.feature_version].add(self.feature_name)
-        self.log_usage_warning(tv)
+        self.log_usage_warning(fname, line_number, tv)
 
     @classmethod
     def report(cls, subproject):
@@ -224,10 +233,12 @@ class FeatureCheckBase:
     def __call__(self, f):
         @wraps(f)
         def wrapped(*wrapped_args, **wrapped_kwargs):
-            subproject = _get_callee_args(wrapped_args, want_subproject=True)[4]
+            callee_args = _get_callee_args(wrapped_args, want_subproject=True)
+            subproject = callee_args[4]
+            fname, lineno = _fname_and_lineno_from_callee_args(callee_args)
             if subproject is None:
                 raise AssertionError('{!r}'.format(wrapped_args))
-            self.use(subproject)
+            self.use(fname, lineno, subproject)
             return f(*wrapped_args, **wrapped_kwargs)
         return wrapped
 
@@ -242,9 +253,10 @@ class FeatureNew(FeatureCheckBase):
     def get_warning_str_prefix(tv):
         return 'Project specifies a minimum meson_version \'{}\' but uses features which were added in newer versions:'.format(tv)
 
-    def log_usage_warning(self, tv):
-        mlog.warning('Project targetting \'{}\' but tried to use feature introduced '
-                     'in \'{}\': {}'.format(tv, self.feature_version, self.feature_name))
+    def log_usage_warning(self, fname, line_number, tv):
+        assert(False)
+        mlog.warning('%s:%d Project targetting \'{}\' but tried to use feature introduced '
+                     'in \'{}\': {}'.format(fname, line_number, tv, self.feature_version, self.feature_name))
 
 class FeatureDeprecated(FeatureCheckBase):
     """Checks for deprecated features"""
@@ -257,10 +269,10 @@ class FeatureDeprecated(FeatureCheckBase):
     def get_warning_str_prefix(tv):
         return 'Deprecated features used:'
 
-    def log_usage_warning(self, tv):
-        mlog.deprecation('Project targetting \'{}\' but tried to use feature '
+    def log_usage_warning(self, fname, line_number, tv):
+        mlog.deprecation('%s:%d Project targetting \'{}\' but tried to use feature '
                          'deprecated since \'{}\': {}'
-                         ''.format(tv, self.feature_version, self.feature_name))
+                         ''.format(fname, line_number, tv, self.feature_version, self.feature_name))
 
 
 class FeatureCheckKwargsBase:
@@ -274,14 +286,16 @@ class FeatureCheckKwargsBase:
         def wrapped(*wrapped_args, **wrapped_kwargs):
             # Which FeatureCheck class to invoke
             FeatureCheckClass = self.feature_check_class
-            kwargs, subproject = _get_callee_args(wrapped_args, want_subproject=True)[3:5]
+            callee_args = _get_callee_args(wrapped_args, want_subproject=True)
+            kwargs, subproject = callee_args[3:5]
+            fname, lineno = _fname_and_lineno_from_callee_args(callee_args)
             if subproject is None:
                 raise AssertionError('{!r}'.format(wrapped_args))
             for arg in self.kwargs:
                 if arg not in kwargs:
                     continue
                 name = arg + ' arg in ' + self.feature_name
-                FeatureCheckClass(name, self.feature_version).use(subproject)
+                FeatureCheckClass(name, self.feature_version).use(fname, lineno, subproject)
             return f(*wrapped_args, **wrapped_kwargs)
         return wrapped
 
diff --git a/mesonbuild/modules/pkgconfig.py b/mesonbuild/modules/pkgconfig.py
index 2f8b5338..8bea8dc3 100644
--- a/mesonbuild/modules/pkgconfig.py
+++ b/mesonbuild/modules/pkgconfig.py
@@ -346,17 +346,19 @@ class PkgConfigModule(ExtensionModule):
                       'subdirs', 'requires', 'requires_private', 'libraries_private',
                       'install_dir', 'extra_cflags', 'variables', 'url', 'd_module_versions'})
     def generate(self, state, args, kwargs):
+        fname = os.path.join(state.subdir, 'meson.build')
+        lineno = state.current_lineno
         if 'variables' in kwargs:
-            FeatureNew('custom pkgconfig variables', '0.41.0').use(state.subproject)
+            FeatureNew('custom pkgconfig variables', '0.41.0').use(fname, lineno, state.subproject)
         default_version = state.project_version['version']
         default_install_dir = None
         default_description = None
         default_name = None
         mainlib = None
         if not args and 'version' not in kwargs:
-            FeatureNew('pkgconfig.generate implicit version keyword', '0.46.0').use(state.subproject)
+            FeatureNew('pkgconfig.generate implicit version keyword', '0.46.0').use(fname, lineno, state.subproject)
         elif len(args) == 1:
-            FeatureNew('pkgconfig.generate optional positional argument', '0.46.0').use(state.subproject)
+            FeatureNew('pkgconfig.generate optional positional argument', '0.46.0').use(fname, lineno, state.subproject)
             mainlib = getattr(args[0], 'held_object', args[0])
             if not isinstance(mainlib, (build.StaticLibrary, build.SharedLibrary)):
                 raise mesonlib.MesonException('Pkgconfig_gen first positional argument must be a library object')
```

**Comment by eli-schwartz** [comment]
I fixed this as part of #9488 by passing node information to feature checks, and passing node information from feature checks to mlog. I guess due to code reorganization it was a lot easier :) it was actually completely free for all feature decorators after updating the decorator itself, which was neat.
